samples/unit_testing/README.md

```diff
@@ -0,0 +1,180 @@
+
+# Unit Testing Guide
+
+This example shows how we can unit test durable function patterns using python unittest patch and mock constructs and some noteworthy mocks.
+
+## Unit Testing Durable HTTP Start Invocation with decorators for EasyAuth
+
+The Durable HTTP Starter is invoked with a `X-MS-CLIENT-PRINCIPAL` in the header of the HTTP request. When configuring EasyAuth, the function needs to be validated against the claims presented. This validation is done via an authorize decorator in this sample.
+
+When making an HTTP request to the service (GET or PUT), you can use the following token to act as both a SubscriptionManager (PUT) and a SubscriptionReader (GET):
+
+`ICAgICAgICB7CiAgICAgICAgICAgICJhdXRoX3R5cCI6ICJhYWQiLAogICAgICAgICAgICAiY2xhaW1zIjogW3sKICAgICAgICAgICAgICAgICJ0eXAiOiAiaHR0cDovL3NjaGVtYXMueG1sc29hcC5vcmcvd3MvMjAwNS8wNS9pZGVudGl0eS9jbGFpbXMvc3VybmFtZSIsCiAgICAgICAgICAgICAgICAidmFsIjogIlVzZXIiCiAgICAgICAgICAgIH0sIHsKICAgICAgICAgICAgICAgICJ0eXAiOiAiZ3JvdXBzIiwKICAgICAgICAgICAgICAgICJ2YWwiOiAiZWY2ZDJkMWEtNzhlYi00YmIxLTk3YzctYmI4YThlNTA5ZTljIgogICAgICAgICAgICB9LCB7CiAgICAgICAgICAgICAgICAidHlwIjogImdyb3VwcyIsCiAgICAgICAgICAgICAgICAidmFsIjogIjNiMjMxY2UxLTI5YzEtNDQxZS1iZGRiLTAzM2Y5NjQwMTg4OCIKICAgICAgICAgICAgfSwgewogICAgICAgICAgICAgICAgInR5cCI6ICJuYW1lIiwKICAgICAgICAgICAgICAgICJ2YWwiOiAiVGVzdCBVc2VyIgogICAgICAgICAgICB9XSwKICAgICAgICAgICAgIm5hbWVfdHlwIjogImh0dHA6Ly9zY2hlbWFzLnhtbHNvYXAub3JnL3dzLzIwMDUvMDUvaWRlbnRpdHkvY2xhaW1zL25hbWUiLAogICAgICAgICAgICAicm9sZV90eXAiOiAiaHR0cDovL3NjaGVtYXMubWljcm9zb2Z0LmNvbS93cy8yMDA4LzA2L2lkZW50aXR5L2NsYWltcy9yb2xlIgogICAgICAgIH0=`
+
+For unit testing the decorator (that get's initialized with a specific environment variable) we just patch in the variable before importing the Durable HTTP Start method like this:
+
+```python
+with patch.dict(os.environ={"SecurityGroups_SUBSCRIPTION_MANAGERS":"ef6d2d1a-78eb-4bb1-97c7-bb8a8e509e9c"}):
+    from ..CreateEnvironmentHTTPStart import main
+````
+
+and make sure we are sending the right `X-MS-CLIENT-PRINCIPAL` in the header of the mocked HttpRequest as seen in [this](./subscription-manager/tests/test_createenvhttpstart_validauth.py) test.
+
+
+We are patching the group-id that gets base64 decoded and compared with the above claims principal sent in the header of the http request
+
+Refer [Auth](./subscription-manager/Auth/authorization.py) for details on how this works
+
+---
+
+## Unit Testing Orchestrator Function
+
+When mocking an orchestrator, the durable orchestration context is mocked like this:
+
+```python
+with patch('azure.durable_functions.DurableOrchestrationContext',spec=df.DurableOrchestrationContext) as mock:
+            mock.get_input = MagicMock(return_value={'displayName' : 'test'})
+            mock.call_sub_orchestrator.side_effect = sub_orc_mock
+            mock.task_all.side_effect = task_all_mock
+        
+            # To get generator results do a next. If orchestrator response is a list, then wrap the function call around a list

```

Comment:
Instead of saying next(orchestrator_fn) etc, just do list generator expression, so that it runs through the orchestrator and gives back a list which can be inspected.

Reply:
Ah, I think I understand this now. This is similar to how `map()` returns a lazily-evaluated list, so it requires a call to `list()` in order to obtain the materialized result, isn't it?

In this case, I think it would be simpler if we continued using `next()`, just for consistency :)